### ISSUE TITLE
Rule proposal: check that function is not returning promise and non-promise

### ISSUE THREAD

**Issue by benmccann** [typescript-eslint/typescript-eslint#9296]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

We have a number of public APIs in our project that return things like `string | Promise<string>`. In fact, we have so many of these that we have a helper type called `MaybePromise`. I believe it would be better if these functions always returned a `Promise`. It would make a nicer API for callers because they would only have to worry about handling rejected promise errors and would not have to worry about errors that are thrown as well

I probably wouldn't want to apply this check to our whole project as internally we can have errors propagate in many functions, but I'd love to apply it to our public type definition file (`.d.ts`).

### Fail Cases

```TypeScript
export interface Emulator {
	platform(details: { config: any; prerender: PrerenderOption }): App.Platform | Promise<App.Platform>;
}
```

### Pass Cases

```TypeScript
export interface Emulator {
	platform(details: { config: any; prerender: PrerenderOption }): Promise<App.Platform>;
}
```

### Additional Info

[`promise-function-async`](https://typescript-eslint.io/rules/promise-function-async) has a somewhat similar goal, but doesn't work for me because it ignores methods with explicit return types, which is one of the things I'd like to detect and disallow. It also has a ton of false positives on my existing codebase where it's almost always fine for me to just propagate errors upwards

**Comment by bradzacher** [comment]
You could probably get most of the way there with `no-restricted-syntax` eg [playground](https://typescript-eslint.io/play/#ts=5.4.5&showAST=es&fileType=.tsx&esQuery=N4IgZglgNgLgpgJxALhCAvkA&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgUQFsBXAG0xmgEY4BvAKDjjHJnWkIAoATYGTBKQDOALjpxsEJOgQBzMZiQBPANzMoaYEl5QFyuAF8AlGKEwoyWXAA%2BcAApQIhBEOAAeMxaSyAfCvoG9PSgkLCIKGhYuAQkrNAATHSMzKzsUFy8-IKi4pLScnqq6praaIWGJvaOzq4e5pZ%2BAfRAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaaFAF2gEsBjUxAE0OQE9dSBDAD3TAG1xsciaNHzRIAGn4CsA7JGSIkNUd0gAVAMpqmAB0QBBXAXakK%2BXADoSpWNFza9YAHxhNAVVxn7uxM9dafACVEADMhRFwqXxcASToI0xCKIR5cNgBbXwBeMAByAAURdIoFXIBdC1IfADkMxAkpWUhM5GQ2AHN6jEg6c1IwWAUwHSKS3wpcMDYwa1tJqsdYT3NIRogAX0ay-k31oA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by Josh-Cena** [comment]
Maybe this can be an option of promise-function-async? I also think promise-function-async is generally useful beyond error handling, as it prevents a function to be maybe-synchronous, which is often bad for side effect ordering.

**Comment by bradzacher** [comment]
`promise-function-async` will do this for function implementations if and only if the return type isn't annotated.

But it doesn't do it for function types or method signatures - only implementations.

A separate rule might be good here to just enforce globally for all function cases. It might be good to have two modes - one where it just looks for `| Promise<T>` in a return type and errors (eg the super strict "if you return a promise at all the entire function must only return a promise") and one where its a bit looser and only looks for `T | Promise<T>`.

**Comment by kirkwaiblinger** [comment]
Looks like there's generally positive response from 3 team members, so marking as accepting PRs 👍

**Comment by bradzacher** [comment]
Based on our current expectations for enhancement requests, I believe this request fails the "nicheness" test:
- This request has been open for 2 years and has netted a total of 1 non-team member interactions.
- This request has been open for 2 years and nobody has moved to implement it.
- The request can _mostly_ be covered by a combination  of `promise-function-async` and `no-restricted-syntax`

For these reasons, I believe we should move to reject and close this. Thoughts @typescript-eslint/triage-team?

**Comment by JoshuaKGoldberg** [comment]
Yeah. I remember at the time thinking that this would be a  interesting quirk of design that would be nice to find for people. But since then I don't recall ever seeing it come up.

Agreed that this deserves to live in a more dedicated plugin.
